File: backend/app/routers/units.py
"""Unit API routes."""

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_session
from app.models.faction import Faction
from app.models.item import Equipment, Skill, Weapon
from app.models.unit import Loadout, Profile, Unit
from app.schemas.unit import (
    ItemRef,
    LoadoutResponse,
    ProfileResponse,
    UnitDetailResponse,
    UnitSummaryResponse,
)

router = APIRouter(prefix="/api/units", tags=["units"])


def _resolve_items(
    items_json: list[dict[str, Any]],
    catalog: dict[int, str],
    extras_map: dict[int, str] | None = None,
) -> list[ItemRef]:
    """Resolve JSONB item references to ItemRef with names."""
    result = []
    for item in items_json:
        item_id = item.get("id", 0)
        name = catalog.get(item_id, f"Unknown ({item_id})")
        extra = item.get("extra", [])
        extra_display = []
        if extras_map and extra:
            extra_display = [extras_map.get(e, str(e)) for e in extra]
        result.append(
            ItemRef(id=item_id, name=name, extra=extra, extra_display=extra_display)
        )
    return result


async def _get_catalogs(
    session: AsyncSession,
) -> tuple[dict[int, str], dict[int, str], dict[int, str], dict[int, str]]:
    """Load item catalogs for name resolution."""
    weapons = {w.id: w.name for w in (await session.execute(select(Weapon))).scalars()}
    skills = {s.id: s.name for s in (await session.execute(select(Skill))).scalars()}
    equips = {
        e.id: e.name for e in (await session.execute(select(Equipment))).scalars()
    }
    # TODO: load extras_map from a dedicated table or cache
    return weapons, skills, equips, {}


def _build_profile_response(
    p: Profile,
    weapon_map: dict[int, str],
    skill_map: dict[int, str],
    equip_map: dict[int, str],
    extras_map: dict[int, str],
) -> ProfileResponse:
    return ProfileResponse(
        id=p.id,
        profile_group_id=p.profile_group_id,
        name=p.name,
        mov=f"{p.mov_1}-{p.mov_2}",
        cc=p.cc,
        bs=p.bs,
        ph=p.ph,
        wip=p.wip,
        arm=p.arm,
        bts=p.bts,
        wounds=p.wounds,
        silhouette=p.silhouette,
        is_structure=p.is_structure,
        unit_type=p.unit_type,
        skills=_resolve_items(p.skills_json or [], skill_map, extras_map),
        equipment=_resolve_items(p.equipment_json or [], equip_map, extras_map),
        weapons=_resolve_items(p.weapons_json or [], weapon_map, extras_map),
    )


def _build_loadout_response(
    lo: Loadout,
    weapon_map: dict[int, str],
    skill_map: dict[int, str],
    equip_map: dict[int, str],
    extras_map: dict[int, str],
) -> LoadoutResponse:
    return LoadoutResponse(
        id=lo.id,
        option_id=lo.option_id,
        profile_group_id=lo.profile_group_id,
        name=lo.name,
        points=lo.points,
        swc=lo.swc,
        skills=_resolve_items(lo.skills_json or [], skill_map, extras_map),
        equipment=_resolve_items(lo.equipment_json or [], equip_map, extras_map),
        weapons=_resolve_items(lo.weapons_json or [], weapon_map, extras_map),
    )


@router.get("", response_model=list[UnitSummaryResponse])
async def list_units(
    faction: str | None = Query(None, description="Filter by faction slug"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    session: AsyncSession = Depends(get_session),
) -> list[UnitSummaryResponse]:
    """List units with optional faction filter."""
    query = select(Unit).options(
        selectinload(Unit.factions), selectinload(Unit.loadouts)
    )

    if faction:
        query = query.join(Unit.factions).where(Faction.slug == faction)

    query = query.order_by(Unit.name).limit(limit).offset(offset)
    result = await session.execute(query)
    units = result.scalars().unique().all()

    return [
        UnitSummaryResponse(
            id=u.id,
            isc=u.isc,
            name=u.name,
            slug=u.slug,
            factions=[f.slug for f in u.factions],
            points_min=min((lo.points for lo in u.loadouts), default=0),
            points_max=max((lo.points for lo in u.loadouts), default=0),
        )
        for u in units
    ]


@router.get("/{slug}", response_model=UnitDetailResponse)
async def get_unit(
    slug: str, session: AsyncSession = Depends(get_session)
) -> UnitDetailResponse:
    """Get full unit details with profiles and loadouts."""
    result = await session.execute(
        select(Unit)
        .options(
            selectinload(Unit.factions),
            selectinload(Unit.profiles),
            selectinload(Unit.loadouts),
        )
        .where(Unit.slug == slug)
    )
    unit = result.scalar_one_or_none()
    if not unit:
        raise HTTPException(status_code=404, detail=f"Unit '{slug}' not found")

    weapon_map, skill_map, equip_map, extras_map = await _get_catalogs(session)

    return UnitDetailResponse(
        id=unit.id,
        isc=unit.isc,
        name=unit.name,
        slug=unit.slug,
        factions=[f.slug for f in unit.factions],
        profiles=[
            _build_profile_response(p, weapon_map, skill_map, equip_map, extras_map)
            for p in sorted(unit.profiles, key=lambda p: (p.profile_group_id, p.id))
        ],
        loadouts=[
            _build_loadout_response(lo, weapon_map, skill_map, equip_map, extras_map)
            for lo in sorted(
                unit.loadouts, key=lambda lo: (lo.profile_group_id, lo.option_id)
            )
        ],
    )

File: backend-ts/src/routes/units.ts
import { Hono } from 'hono';
import { and, asc, eq, inArray } from 'drizzle-orm';
import { db } from '../db/client.js';
import {
    units,
    factions,
    unit_factions,
    profiles,
    loadouts,
} from '../db/schema.js';
import { resolveItems, type ItemRef } from '../lib/items.js';
import { getCatalogs } from '../lib/catalogs.js';

const router = new Hono();

interface UnitSummary {
    id: number;
    isc: string;
    name: string;
    slug: string;
    factions: string[];
    points_min: number;
    points_max: number;
}

router.get('/', async c => {
    const factionSlug = c.req.query('faction') ?? undefined;
    const limit = Math.min(Math.max(parseInt(c.req.query('limit') ?? '50', 10) || 50, 1), 200);
    const offset = Math.max(parseInt(c.req.query('offset') ?? '0', 10) || 0, 0);

    let unitIds: number[] | null = null;
    if (factionSlug) {
        const matched = await db
            .select({ unit_id: unit_factions.unit_id })
            .from(unit_factions)
            .innerJoin(factions, eq(factions.id, unit_factions.faction_id))
            .where(eq(factions.slug, factionSlug));
        unitIds = matched.map(r => r.unit_id);
        if (unitIds.length === 0) return c.json([]);
    }

    const baseRows = await db
        .select({
            id: units.id,
            isc: units.isc,
            name: units.name,
            slug: units.slug,
        })
        .from(units)
        .where(unitIds ? inArray(units.id, unitIds) : undefined)
        .orderBy(asc(units.name))
        .limit(limit)
        .offset(offset);

    if (baseRows.length === 0) return c.json([]);
    const ids = baseRows.map(u => u.id);

    const factionRows = await db
        .select({
            unit_id: unit_factions.unit_id,
            slug: factions.slug,
        })
        .from(unit_factions)
        .innerJoin(factions, eq(factions.id, unit_factions.faction_id))
        .where(inArray(unit_factions.unit_id, ids));

    const loadoutRows = await db
        .select({ unit_id: loadouts.unit_id, points: loadouts.points })
        .from(loadouts)
        .where(inArray(loadouts.unit_id, ids));

    const factionsByUnit = new Map<number, string[]>();
    for (const r of factionRows) {
        const arr = factionsByUnit.get(r.unit_id) ?? [];
        arr.push(r.slug);
        factionsByUnit.set(r.unit_id, arr);
    }
    const ptsByUnit = new Map<number, number[]>();
    for (const r of loadoutRows) {
        const arr = ptsByUnit.get(r.unit_id) ?? [];
        arr.push(r.points);
        ptsByUnit.set(r.unit_id, arr);
    }

    const result: UnitSummary[] = baseRows.map(u => {
        const pts = ptsByUnit.get(u.id) ?? [];
        return {
            id: u.id,
            isc: u.isc,
            name: u.name,
            slug: u.slug,
            factions: factionsByUnit.get(u.id) ?? [],
            points_min: pts.length ? Math.min(...pts) : 0,
            points_max: pts.length ? Math.max(...pts) : 0,
        };
    });

    return c.json(result);
});

interface ProfileResponse {
    id: number;
    profile_group_id: number;
    name: string;
    mov: string;
    cc: number;
    bs: number;
    ph: number;
    wip: number;
    arm: number;
    bts: number;
    wounds: number;
    silhouette: number;
    is_structure: boolean;
    unit_type: number | null;
    skills: ItemRef[];
    equipment: ItemRef[];
    weapons: ItemRef[];
}

interface LoadoutResponse {
    id: number;
    option_id: number;
    profile_group_id: number;
    name: string;
    points: number;
    swc: number;
    skills: ItemRef[];
    equipment: ItemRef[];
    weapons: ItemRef[];
}

interface UnitDetailResponse {
    id: number;
    isc: string;
    name: string;
    slug: string;
    factions: string[];
    profiles: ProfileResponse[];
    loadouts: LoadoutResponse[];
}

router.get('/:slug', async c => {
    const slug = c.req.param('slug');

    const [unit] = await db.select().from(units).where(eq(units.slug, slug)).limit(1);
    if (!unit) return c.json({ detail: `Unit '${slug}' not found` }, 404);

    const [factionRows, profileRows, loadoutRows, catalogs] = await Promise.all([
        db
            .select({ slug: factions.slug })
            .from(unit_factions)
            .innerJoin(factions, eq(factions.id, unit_factions.faction_id))
            .where(eq(unit_factions.unit_id, unit.id)),
        db.select().from(profiles).where(eq(profiles.unit_id, unit.id)),
        db.select().from(loadouts).where(eq(loadouts.unit_id, unit.id)),
        getCatalogs(),
    ]);

    const sortedProfiles = [...profileRows].sort(
        (a, b) => a.profile_group_id - b.profile_group_id || a.id - b.id,
    );
    const sortedLoadouts = [...loadoutRows].sort(
        (a, b) => a.profile_group_id - b.profile_group_id || a.option_id - b.option_id,
    );

    const response: UnitDetailResponse = {
        id: unit.id,
        isc: unit.isc,
        name: unit.name,
        slug: unit.slug,
        factions: factionRows.map(f => f.slug),
        profiles: sortedProfiles.map(p => ({
            id: p.id,
            profile_group_id: p.profile_group_id,
            name: p.name,
            mov: `${p.mov_1}-${p.mov_2}`,
            cc: p.cc,
            bs: p.bs,
            ph: p.ph,
            wip: p.wip,
            arm: p.arm,
            bts: p.bts,
            wounds: p.wounds,
            silhouette: p.silhouette,
            is_structure: p.is_structure,
            unit_type: p.unit_type,
            skills: resolveItems(p.skills_json, catalogs.skills),
            equipment: resolveItems(p.equipment_json, catalogs.equipment),
            weapons: resolveItems(p.weapons_json, catalogs.weapons),
        })),
        loadouts: sortedLoadouts.map(lo => ({
            id: lo.id,
            option_id: lo.option_id,
            profile_group_id: lo.profile_group_id,
            name: lo.name,
            points: lo.points,
            swc: lo.swc,
            skills: resolveItems(lo.skills_json, catalogs.skills),
            equipment: resolveItems(lo.equipment_json, catalogs.equipment),
            weapons: resolveItems(lo.weapons_json, catalogs.weapons),
        })),
    };

    return c.json(response);
});

export default router;
// Suppress unused warnings for `and` in case future filters need it
void and;
